### PR TITLE
Add overloads for sum

### PR DESCRIPTION
**Pull request by jmh530**
This PR adds the overloads for `sum` that exist for `prod` and some of the other statistical functions already.

### CODE REVIEW

File: source/mir/math/sum.d [hunk 1]
```diff
@@ -1722,6 +1750,15 @@ template sum(Summation summation = Summation.appropriate)
         import core.lifetime: move;
         return .sum!(F, ResolveSummationType!(summation, Range, F))(r.move, seed);
     }
+
+    ///
+    sumType!(CommonType!T) sum(T...)(T r)
```
Comment (9il) [inline source/mir/math/sum.d:1755]: Why this overload is required?

**Comment by jmh530** [comment]
I added two overloads as part of this PR. This one is for when you do
`sum(1, 2, 3)` and also `sum(1f, 2, 3)`. The other one is for when you do
`sum!float(1, 2, 3)` or `sum!float(1f, 2f, 3f)`. The main difference is
that it means the user does not have to provide the type.

 I wouldn't say it is required. It is a quality-of-life enhancement. The
ability to call functions like `f(1, 2, 3)` and have the same behavior as
`f([1, 2, 3])` is part of D and I see no reason not to incorporate it if
someone is willing to put in the work to add them. If they should not be
added for `sum`, then they should be removed from other functions that I
have added them for consistency. I think it is valuable to have this
functionality for some of these numerical functions, but I would understand
if you were not in favor of that.

**Comment by 9il** [comment]
> I added two overloads as part of this PR. This one is for when you do `sum(1, 2, 3)` and also `sum(1f, 2, 3)`. The other one is for when you do `sum!float(1, 2, 3)` or `sum!float(1f, 2f, 3f)`. The main difference is that it means the user does not have to provide the type.

The following works well without a variadic template. It is very preferred to avoid variadic templates of functions, the generates a lot of template bloat and hard to maintain.

```
F sum(F)(scope F[] ar...)
{
    return ar[0];
}
void main() {
    auto s = sum(1f, 2, 3);
}
```

**Comment by jmh530** [comment]
@9il Sounds good. I used `scope const` instead of just `scope` since you had recommended that previously. 

One question I had as I was working on this is that I did not use `core.lifetime.move` for the typesafe variadic functions. Is this ok? I don't have a good sense of when that should be used (definitely with slice/range but otherwise?).

**Comment by 9il** [comment]
> One question I had as I was working on this is that I did not use `core.lifetime.move` for the typesafe variadic functions. Is this ok? I don't have a good sense of when that should be used (definitely with slice/range but otherwise?).

Yes. But it would make sense to add `move` to `fold` if it hasn't been used before.

**Pull request merged**